In [ ]:
!pip install -q pyomo
from pyomo.environ import *

     |████████████████████████████████| 9.6 MB 5.2 MB/s 
     |████████████████████████████████| 49 kB 5.2 MB/s 


In [ ]:
import numpy as np

In [ ]:
#Creating a model
model1 = ConcreteModel()

#No. of decision variables
N = 12
M = 5

#Objective function coefficient using numpy array
obj_coef_c = np.array([0.56, 1.721, -9.6, -2.85, 41.5, 33.05, -0.75, 20.5, -0.35, -31.02, 50.65, 6.125])

#Constraints coeffiecient using numpy array
constr_coef_A = np.array([1, -1, 0, 0, 2, -2, 1, 0, 0, -1, 1, -2, 1, 0, 0, -1, 0, -2, 0, -1, 1, 0, 0, -1, -1, 0, 0, 1, -1, 2.5, -3.5, 0, 2, -1, 1, 0, 0, 1, -0.45, 0, -2.78, 0, -3.4, 0, -1, 1, -1, 0.25, 0, -1, 1, 0, 1, 0, 1, 3, -2, 0, 1, 0]).reshape(5, 12)

In [ ]:
#Declaring RHS of the constraints as variables
constr1_rhs_b = np.array([88.5, 35.4, 47.9, 41.43, 22.7])

#Declaring lower and upper bounds of the constraints
l = np.array([0, 0, -np.inf, -np.inf, -34, 0, -67, 0, -23, -5, 6, 1])
u = np.array([12, 503, 4, +np.inf, +np.inf, 2, 20, 55, 23, 15, +np.inf, +np.inf])

#Set of row and column indices
row_indices = np.arange(M)
col_indices = np.arange(N)
print(col_indices)
print(col_indices[1])

[ 0  1  2  3  4  5  6  7  8  9 10 11]
1


In [ ]:
#Declaring the decision variables in model
model1.y = Var(col_indices)

In [ ]:
#Creating a constraint list to hold multiple constraints
model1.constraints = ConstraintList()

#Adding model constraints one by one to the list
#add the constraint p1x1 + p2x2 +.......+ pNxN == b1
for i in row_indices:
 model1.constraints.add(sum(constr_coef_A[i][j]*model1.y[j] for j in col_indices) <= constr1_rhs_b[i])

In [ ]:
#add the upper and lower bounds for the variables
for j in col_indices:
  model1.y[j].setlb(l[j])
  model1.y[j].setub(u[j])

In [ ]:
#add the model objective
model1.objective = Objective(expr = sum(obj_coef_c[j]*model1.y[j] for j in col_indices), sense = minimize)

In [ ]:
#print the whole model
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    y_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    y : Size=12, Index=y_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :   0.0 :  None :  12.0 : False :  True :  Reals
          1 :   0.0 :  None : 503.0 : False :  True :  Reals
          2 :  None :  None :   4.0 : False :  True :  Reals
          3 :  None :  None :  None : False :  True :  Reals
          4 : -34.0 :  None :  None : False :  True :  Reals
          5 :   0.0 :  None :   2.0 : False :  True :  Reals
          6 : -67.0 :  None :  20.0 : False :  True :  Reals
          7 :   0.0 :  None :  55.0 : False :  True :  Reals
          8 : -23.0 :  None :  23.0 : False 

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -1951.831
  Upper bound: -1951.831
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 13
  Number of nonzeros: 38
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01181173324584961
# ----------------------------------------------------------
#   Solution Information
# ---------------------------

In [ ]:
#Display solution
print('\nObjective = ', model1.objective())
print('\nDecision Variables')
for i in col_indices:
  print('y[',i,'] = ', model1.y[i].value)
print('\nConstraints')
model1.constraints.display()


Objective =  -1951.8309999999997

Decision Variables
y[ 0 ] =  12.0
y[ 1 ] =  0.0
y[ 2 ] =  4.0
y[ 3 ] =  119.82
y[ 4 ] =  -34.0
y[ 5 ] =  0.0
y[ 6 ] =  20.0
y[ 7 ] =  0.0
y[ 8 ] =  -7.46000000000002
y[ 9 ] =  15.0
y[ 10 ] =  6.0
y[ 11 ] =  1.0

Constraints
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :               -47.0 :  88.5
      2 :  None : -116.28000000000002 :  35.4
      3 :  None :   47.89999999999995 :  47.9
      4 :  None :   41.43000000000002 : 41.43
      5 :  None :   10.92000000000004 :  22.7


**Report on exercise 2 part 1:**

*   **The value of the decision variables at optimal point are:**

     y1 = 12.0

     y2 = 0.0

     y3 = 4.0

     y4 = 119.82

     y5 = -34.0

     y6 = 0.0

     y7 = 20.0

     y8 = 0.0

     y9 = -7.46000000000002

     y10 = 15.0

     y11 = 6.0

     y12 = 1.0

*   **The value of the objective function at the optimal point is:**

     -1951.8309999999997


    



**Report on exercise 2 part 2:**



*   Constraint 1 is an inactive constraint because it evaluates to -47.0 < 88.5 = RHS
*   Constraint 2 is an inactive constraint because it evaluates to -116.28000000000002 < 35.4= RHS

*   Constraint 3 is an active constraint because it evaluates to 47.8999999999 which is almost equal to 47.9 = RHS

*   Constraint 4 is an active constraint because it evaluates to 41.43 which is almost equal to 41.43 = RHS

*   Constraint 2 is an inactive constraint because it evaluates to 10.92 > 22.7 = RHS






